spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/ChangelogRowReader.java

```diff
@@ -0,0 +1,129 @@
+      boolean caseSensitive) {
+    super(table, taskGroup, expectedSchema, caseSensitive);
+  }
+
+  @Override
+  protected CloseableIterator<InternalRow> open(ChangelogScanTask task) {
+    if (task instanceof AddedRowsScanTask) {
+      return openAddedRowsScanTask((AddedRowsScanTask) task);
+    } else if (task instanceof DeletedRowsScanTask) {
+      throw new UnsupportedOperationException("Deleted rows scan task is not supported yet");
+    } else if (task instanceof DeletedDataFileScanTask) {
+      return openDeletedDataFileScanTask((DeletedDataFileScanTask) task);
+    } else {
+      throw new IllegalArgumentException(
+          "Unsupported changelog scan task type: " + task.getClass().getName());
+    }
+  }
+
+  @Override
+  protected Stream<ContentFile<?>> referencedFiles(ChangelogScanTask task) {
+    if (task instanceof AddedRowsScanTask) {
+      return Stream.concat(
+          Stream.of(((AddedRowsScanTask) task).file()),
+          ((AddedRowsScanTask) task).deletes().stream());
+    } else if (task instanceof DeletedRowsScanTask) {
+      throw new UnsupportedOperationException("Deleted rows scan task is not supported yet");
+    } else if (task instanceof DeletedDataFileScanTask) {
+      return Stream.concat(
+          Stream.of(((DeletedDataFileScanTask) task).file()),
+          ((DeletedDataFileScanTask) task).existingDeletes().stream());
+    } else {
+      throw new IllegalArgumentException(
+          "Unsupported changelog scan task type: " + task.getClass().getName());
+    }
+  }
+
+  CloseableIterator<InternalRow> openAddedRowsScanTask(AddedRowsScanTask task) {
+    SparkDeleteFilter deletes =
+        new SparkDeleteFilter(task.file().path().toString(), task.deletes());
+    return deletes.filter(internalRowIterable(task, deletes.requiredSchema())).iterator();
+  }
+
+  private CloseableIterator<InternalRow> openDeletedDataFileScanTask(DeletedDataFileScanTask task) {
+    SparkDeleteFilter deletes =
+        new SparkDeleteFilter(task.file().path().toString(), task.existingDeletes());
+    return deletes.filter(internalRowIterable(task, deletes.requiredSchema())).iterator();
+  }
+
+  private CloseableIterable<InternalRow> internalRowIterable(
+      ContentScanTask<DataFile> task, Schema readSchema) {
+    // schema or rows returned by readers
+    Map<Integer, ?> idToConstant = constantsMap(task, readSchema);
+
+    String filePath = task.file().path().toString();
+
+    // update the current file for Spark's filename() function
+    InputFileBlockHolder.set(filePath, task.start(), task.length());
+
+    InputFile location = getInputFile(filePath);
+    Preconditions.checkNotNull(location, "Could not find InputFile");
+    return newIterable(
+        location,
+        task.file().format(),
+        task.start(),
+        task.length(),
+        task.residual(),
+        readSchema,
+        idToConstant);
+  }
+
+  @Override
+  public InternalRow get() {
+    JoinedRow cdcRow = new JoinedRow();
```

Comment:
Instead of allocating a new joined and metadata row for every data row, I'd rather define them once and reuse by setting the row in the iterator. We can init these helper rows once per task.

```
JoinedRow cdcRow = new JoinedRow();

InternalRow metadata = ...;

cdcRow.withRight(metadata);

CloseableIterable<InternalRow> cdcRows = CloseableIterable.transform(deletes.filter(...), cdcRow::withLeft);
return cdcRows.iterator();
```